GridDist and SuperCellSize options for PICMI

pordyna opened the pull request:
As in the title. @chillenzer can you have a look?

chillenzer commented on test/python/picongpu/quick/picmi/grid.py:
```diff
@@ -84,3 +84,120 @@ def test_n_gpus_wrong_numbers(self):
                 grid.get_as_pypicongpu()
+
+    def test_supercell(self):
+        """test supercell default"""
```
This doesn't quite sound like it's an accurate description. Please choose one of the following:
1. be more verbose and explicit such that I can match what the test is doing with the description
2. or choose another description that better reflects that.

chillenzer commented on test/python/picongpu/quick/picmi/grid.py:
```diff
@@ -84,3 +84,120 @@ def test_n_gpus_wrong_numbers(self):
+        assert g.super_cell_size == (8, 8, 4), "supercell should be [8,8,4]"
+
+    def test_super_cell_mismatch_no_dist(self):
+        with self.assertRaisesRegex(Exception, ".*GPU- and/or super-cell-distribution.*"):
+            grid = picmi.Cartesian3DGrid(
+                number_of_cells=[192, 2048, 12],
+                lower_bound=[0, 0, 0],
+                upper_bound=[3.40992e-5, 9.07264e-5, 2.1312e-6],
+                lower_boundary_conditions=["open", "open", "periodic"],
+                upper_boundary_conditions=["open", "open", "periodic"],
+                picongpu_super_cell_size=(7, 8, 4),
+            )
```
Untested but you get the idea:
```suggestion
        number_of_cells = [192, 2048, 12]
        super_cell_size = [7, 8, 4]
        # pre-condition for a mismatch
        assert any(np.asarray(number_of_cells) % super_cell_size != 0)
        with self.assertRaisesRegex(Exception, ".*GPU- and/or super-cell-distribution.*"):
            grid = picmi.Cartesian3DGrid(
                number_of_cells=number_of_cells,
                lower_bound=[0, 0, 0],
                upper_bound=[3.40992e-5, 9.07264e-5, 2.1312e-6],
                lower_boundary_conditions=["open", "open", "periodic"],
                upper_boundary_conditions=["open", "open", "periodic"],
                picongpu_super_cell_size=super_cell_size,
            )
```

This is non-blocking and maybe personal style but I like to be explicit about what's the aspects to focus on in a test. If you like this approach, you'll surely figure out how to apply it to the other tests yourself.

pordyna replied:
You will have to take your beef to ChatGPT, I was 100% vibe coding the tests :D, but I agree hahaha

pordyna replied:
I will allow myself to skip this for now :D

chillenzer replied:
Okay.

chillenzer commented on test/python/picongpu/quick/picmi/grid.py:
```diff
@@ -84,3 +84,120 @@ def test_n_gpus_wrong_numbers(self):
+            lower_bound=[0, 0, 0],
+            upper_bound=[3.40992e-5, 9.07264e-5, 2.1312e-6],
+            lower_boundary_conditions=["open", "open", "periodic"],
+            upper_boundary_conditions=["open", "open", "periodic"],
```
There's a number of `kwargs` that never change. I'd suggest that you define a variable `UNIMPORTANT_KWARGS` and use `**UNIMPORTANT_KWARGS` to more clearly focus the reader on the important aspects of the test. You can probably come up with a better name for this variable.

chillenzer commented on test/python/picongpu/quick/picmi/grid.py:
```diff
@@ -84,3 +84,120 @@ def test_n_gpus_wrong_numbers(self):
+                lower_boundary_conditions=["open", "open", "periodic"],
+                upper_boundary_conditions=["open", "open", "periodic"],
+                picongpu_n_gpus=[1, 1, 1],
+                picongpu_grid_dist=([96, 96], [2048], [12]),  # length 2 in x but n_gpus=1
```
```suggestion
                # length 2 in x but n_gpus=1
                picongpu_grid_dist=([96, 96], [2048], [12]), 
```

chillenzer commented on test/python/picongpu/quick/picmi/grid.py:
```diff
@@ -84,3 +84,120 @@ def test_n_gpus_wrong_numbers(self):
+            )
+            grid.get_as_pypicongpu()
+
+    def test_grid_dist_correct(self):
```
After all those tests about failures it's reassuring to know that there actually exists a valid input. :wink:

chillenzer commented on test/python/picongpu/quick/picmi/grid.py:
```diff
@@ -84,3 +84,120 @@ def test_n_gpus_wrong_numbers(self):
+                picongpu_super_cell_size=(7, 8, 4),
+                picongpu_grid_dist=([12, 180], [2048], [12]),
+            )
+            grid.get_as_pypicongpu()
```
Also: Please be explicit about which call is the raising one. Does this get checked in the constructor or in the `get_as_pypicongpu()` call. Only guard that single raising statement.

chillenzer commented on share/picongpu/pypicongpu/schema/grid.Grid3D.json:
```diff
@@ -94,12 +117,49 @@
+          "$ref": "#grid_dist_component"
+        },
+        "z": {
+          "$ref": "#grid_dist_component"
```
Oh, how great! Finally someone took the time to look up how this internal referencing works! Thanks!

pordyna replied:
We love copy-pasting stuff from somewhere else without understanding and hoping it works

chillenzer commented on docs/source/usage/picmi/intro.rst:
```diff
@@ -173,7 +173,18 @@ Parameters/Methods prefixed with ``picongpu_`` are PIConGPU-exclusive.
+  - ``picongpu_grid_dist``: 
+    Optional static grid distrbution. If not set or equal ``None`` the cells are distributed evenly over the GPUs.
+    Tuple with 3 lists of integers or ``None``.
+    Each lists describes the distribution of the grid on the GPUs in the respective direction.
```
```suggestion
    Each list describes the distribution of the grid on the GPUs in the respective direction.
```

chillenzer commented on docs/source/usage/picmi/intro.rst:
```diff
@@ -173,7 +173,18 @@ Parameters/Methods prefixed with ``picongpu_`` are PIConGPU-exclusive.
+  - ``picongpu_super_cell_size``:
+    A tuple with 3 integers, describing the size of a super cell.
+    The default is ``(8, 8, 4)``.
+    It is recomended to set it to ``(16, 16, 1)`` for 2D simulations.
```
2D is not yet supported, technically speaking.

pordyna replied:
yeah but one can already easily run 2D sims with picmi, even with the 3D Grid object. Just change dimension.param

chillenzer replied:
Okay.

chillenzer commented on lib/python/picongpu/picmi/grid.py:
```diff
@@ -90,20 +99,41 @@ def get_as_pypicongpu(self):
+                    len(self.picongpu_grid_dist[i]) == g.n_gpus[i]
+                ), f"number of grid distributions in dimension {i} must match number of gpus"
+        for i in range(3):
+            assert self.picongpu_super_cell_size[i] >= 1, "super cell size must be an integer greater than 1"
```
Not blocking but as a side remark: Assertions in production code are supposed to have a very narrow use case because they can be turned off
```bash
$ python --help
[...]
-O     : remove assert and __debug__-dependent statements; add .opt-1 before
         .pyc extension; also PYTHONOPTIMIZE=x
-OO    : do -O changes and also discard docstrings; add .opt-2 before
         .pyc extension
[...]
```
So for any checks that are supposed to unconditionally run, you'd want to use real `if ...: raise ...` over assertions. Only for the very narrow use case where you're actually checking for something mildly relevant that you explicitly WANT to be able to turn off `assert` should be used. (That doesn't apply to tests, of course.)

Also, in many situations you can be more precise and expressive with raising (custom) exceptions over a general `AssertionError`.

chillenzer commented on lib/python/picongpu/pypicongpu/grid.py:
```diff
@@ -101,4 +111,19 @@ def _get_serialized(self) -> dict:
+        if self.grid_dist is not None:
+            result_dict |= {
+                "grid_dist": {
+                    "x": list(map(lambda x: {"device_cells": x}, self.grid_dist[0])),
```
Non-blocking but some style guides would advice you to prefer
```suggestion
                    "x": [{"device_cells": x} for x in self.grid_dist[0]]
```

You could even reduce the whole to
```python
{direction: [{"device_cells": x} for x in dist] for direction, dist in zip("xyz", self.grid_dist)}
```
This general form could also be helpful once we're allowing 2D.

chillenzer reviewed (dismissed):
Looks good. Nothing blocking.

By the way, beautiful [bikeshed](https://en.wiktionary.org/wiki/bikeshedding) you've got there! I've recently heard that black and white stripes reduce the [the likelihood of getting bitten by some nasty bugs](https://www.bbc.com/future/article/20191031-the-truth-behind-why-zebras-have-stripes). Ever thought about repainting it?

If so, comment or push updates until lunch. If not, I'll merge in the afternoon.

chillenzer commented:
Also... fix the CI, of course!

pordyna commented:
I forgot to actually put gridDist in program params in the cfg...

pordyna committed:
implement review comments

fix missing grod dist in program options

fix ci

fix ci again

fix merge errors

pordyna committed:
add picmi support for super cell size and arbitrary grid dist

add extra checks and tests for picmi gridDist and superCell

picadd grid_Dist and superCell documentation for picmi

pordyna commented:
In case it passes now, the changes since the last review were squished into the last commit

chillenzer reviewed (approved):
Great! Thanks! Particularly for taking the time to rework the tests a little!

There are teeny-tiny things like that `self.COMMON_KWARGS` should be snake-case if member of a class but that's all fine for now.
